packages/ui/src/components/sidebar/sidebar-context.tsx

```diff
@@ -0,0 +1,111 @@
+import { ComponentProps, createContext, forwardRef, useCallback, useContext, useEffect, useMemo, useState } from 'react'
+
+import { cn } from '@utils/cn'
+
+import {
+  SIDEBAR_COLLAPSED_WIDTH,
+  SIDEBAR_COOKIE_MAX_AGE,
+  SIDEBAR_COOKIE_NAME,
+  SIDEBAR_KEYBOARD_SHORTCUT,
+  SIDEBAR_WIDTH
+} from './sidebar-constants'
+import { useIsMobile } from './use-is-mobile'
+
+type SidebarContextType = {
+  state: 'expanded' | 'collapsed'
+  setOpen: (open: boolean) => void
+  openMobile: boolean
+  setOpenMobile: (open: boolean) => void
+  isMobile: boolean
+  toggleSidebar: () => void
+}
+
+const SidebarContext = createContext<SidebarContextType | null>(null)
+
+export type UseSidebarSignature = () => SidebarContextType
+
+export function useSidebar() {
+  const context = useContext(SidebarContext)
+  if (!context) {
+    throw new Error('useSidebar must be used within a SidebarProvider.')
+  }
+
+  return context
+}
+
+export const SidebarProvider = forwardRef<
+  HTMLDivElement,
+  ComponentProps<'div'> & { defaultOpen?: boolean; open?: boolean; onOpenChange?: (open: boolean) => void }
+>(({ defaultOpen = true, open: openProp, onOpenChange: setOpenProp, className, children, ...props }, ref) => {
+  const isMobile = useIsMobile()
+  const [openMobile, setOpenMobile] = useState(false)
+
+  // This is the internal state of the sidebar.
+  // We use openProp and setOpenProp for control from outside the component.
+  const [_open, _setOpen] = useState(defaultOpen)
+  const open = openProp ?? _open
+  const setOpen = useCallback(
+    (value: boolean | ((value: boolean) => boolean)) => {
+      const openState = typeof value === 'function' ? value(open) : value
+      if (setOpenProp) {
+        setOpenProp(openState)
+      } else {
+        _setOpen(openState)
+      }
+
+      // This sets the cookie to keep the sidebar state.
+      document.cookie = `${SIDEBAR_COOKIE_NAME}=${openState}; path=/; max-age=${SIDEBAR_COOKIE_MAX_AGE}`
```

Comment:
What are the benefits of storing the status in cookie than localStorage? Where are we reading its value to persist the sidebar state?

Reply:
This thing has been there since the sidebar was set up for the first time. It probably was generated by the Shadcn. Currently, it's not used anywhere

<img width="895" height="416" alt="image" src="https://github.com/user-attachments/assets/de0085b9-ac19-407c-b5fd-616c5e46f192" />

Reply:
Got it 👍 ... can we add a logic to persist sidebar state of user using localStorage? You can take it as a separate PR.